File: week07/identify_poet.pl
#!/usr/bin/perl
use strict;
use warnings;

use English;

require "total_words.pl";
require "frequency.pl";

my %frequencies = loadPoetFrequencies(1);

foreach my $file (@ARGV) {
    my %probabilities;
    foreach my $poet (keys %frequencies) {
        $probabilities{$poet} = 0;
    }

    open my $input, "<", $file;
    foreach my $word (findWords($input)) {
        $word = lc $word;
        foreach my $poet (keys %frequencies) {
            $probabilities{$poet} += $frequencies{$poet}{$word} || $frequencies{$poet}{"_zero"};
        }
    }

    my $result = (sort {$probabilities{$b} <=> $probabilities{$a}} keys %probabilities)[0];
    printf("%s most resembles the work of %s (log-probability=%.1f)\n", $file, $result, $probabilities{$result});
}

File: week07/identify_poet.py
#!/usr/bin/python3

import io, sys
from total_words import findWords
from frequency import loadPoetFrequencies

frequencies = loadPoetFrequencies(True)

for file in sys.argv[1:]:
    probabilities = {}
    for poet in frequencies.keys():
        probabilities[poet] = 0

    with io.open(file) as input:
        for word in findWords(input):
            word = word.lower()
            for poet, wordFrequencies in frequencies.items():
                probabilities[poet] += wordFrequencies.get(word, wordFrequencies["_zero"])

    result = sorted(probabilities.items(), key = lambda p: p[1])[-1]
    print("{} most resembles the work of {} (log-probability={:.1f})".format(file, result[0], result[1]))
